In [91]:
import pandas as pd
import numpy as np
import math as math
import matplotlib.pyplot as plt
import seaborn as sns
import random
%matplotlib inline 

In [92]:
# to see all the columns when using head() or describe()
pd.set_option('display.max_columns',60)

In [93]:
df = pd.read_csv("kenya_data/diaries_transactions_all.csv", dtype={'account_startclose_balance': str})

## For now I keep only "business and agriculture loans", "joint liability loans", "personal loans", "borrowing from an informal group", "borrowing from family and friends", "shylock loans".

In [95]:
dfb = df[((df["trx_family_code"]=="FRMLN")&((df["trx_type_code"]==2760)|(df["trx_type_code"]==3247)|(df["trx_type_code"]==2762)))|((df["trx_family_code"]=="INFGRP")&(df["trx_type_code"]==3395))|((df["trx_family_code"]=="INFP2P")&((df["trx_type_code"]==2752)|(df["trx_type_code"]==3246)))]

# 1. Get rid of loans of with unexplained balance adjustments

In [96]:
dfb["trx_prx_purpose"].unique()

array(['7. Closing Balance--End of last DQ',
       '6. Closing Balance--End of last DQ',
       '1. Starting balance (today)', '5. Interest accruing',
       'DECREASE--Unexplained balance adjustment',
       'INCREASE--Unexplained balance adjustment', '4. Any known fees',
       '3. Payments', '6. Loan written off', '6. Debt written off',
       '2. New borrowinng (individual portion)', '2. New borrowing',
       '7. Refund from lender'], dtype=object)

We insert a new column that is 0 if there is no unexplained balance adjustment for this account and a 1 if there is one. The column in 1 or 0 for all entries that correspond to that account.

In [97]:
dfb.insert(dfb.shape[1], "acc_unexplained_trx", np.zeros(dfb.shape[0]))

This program goes through all unique accounts, checks whether it has an unexplained balance adjustment and if yes sets the variable u to 1. Then it goes through another for-loop and sets the entries of that column 0 for all accounts that have an unexplained balance adjustment.

In [98]:
for acc in dfb["account_ids"].unique():
    u = 0
    for index,row in dfb[dfb["account_ids"]==acc].iterrows():
        if row["trx_prx_purpose"]=="DECREASE--Unexplained balance adjustment" or row["trx_prx_purpose"]=="INCREASE--Unexplained balance adjustment":
            u = 1
    for index,row in dfb[dfb["account_ids"]==acc].iterrows():
        dfb.at[index,"acc_unexplained_trx"] = u

### Let us check if it worked.

In [99]:
s = set()
for acc in dfb[dfb["acc_unexplained_trx"]==0]["account_ids"].unique():
     for i in dfb[dfb["account_ids"]==acc]["trx_prx_purpose"].unique():
            s.add(i)
print(s)

{'3. Payments', '6. Loan written off', '2. New borrowinng (individual portion)', '7. Closing Balance--End of last DQ', '1. Starting balance (today)', '4. Any known fees', '7. Refund from lender', '5. Interest accruing', '6. Debt written off', '6. Closing Balance--End of last DQ', '2. New borrowing'}


### "refund from lender" and "Debt written off" might be hard to deal with, but there are only 3 of those

In [103]:
dfb[dfb["trx_prx_purpose"]=="6. Debt written off"]

hh_ids  unique_hhs first_trx_date_hh last_trx_date_hh  \
3372  KVIHC22         NaN         01oct2012        30sep2013   

      tot_hh_daysofobs  tot_hh_monthsofobs interview_designation   int_date  \
3372               364                  12  04=Diaries Interview  25sep2013   

      int_month  int_year int_yr_mo first_int_date        account_ids  \
3372          9      2013   2013_09      27aug2012  59134726285800000   

      unique_accnts        m_ids_owner  unique_hm_owner account_bsheet_desig  \
3372            NaN  59134362693600000              NaN            Liability   

     account_startclose_balance account_formal account_liquid  \
3372                        NaN       Informal            NaN   

     first_trx_date_acc last_trx_date_acc  tot_acc_daysofobs  \
3372          22oct2012         25sep2013                338   

      tot_acc_monthsofobs             trx_id          m_ids_trx   trx_date  \
3372                   11  60138624703300000  59134362693600000  01sep2013   

      trx_month  trx_year trx_yr_mo  trx_dq_round  trx_stdtime_days_hh  \
3372          9      2013   2013_09          19.0                  335   

      trx_stdtime_mnths_hh  trx_stdtime_days_acc  trx_stdtime_mnths_acc  \
3372                    11                   314                     10   

     trx_class_code                                     trx_class_desc  \
3372         findev  Borrowing, lending, savings or insurance media...   

     trx_family_code trx_family_desc  trx_type_code  \
3372          INFP2P    Informal P2P           3246   

                         trx_type_desc      trx_prx_purpose  \
3372  Moneylender or Shylock Borrowing  6. Debt written off   

     trx_prx_purpose_fd  trx_fee trx_bsheet_direction  trx_mode_code  \
3372                NaN      0.0             Decrease            1.0   

     trx_mode_desc  trx_place_incommunity  trx_distance_km trx_outlet  \
3372          Cash                    NaN              NaN        NaN   

     trx_direction  trx_value_kes  trx_value_usd  ddd_gift trx_inkind_units  \
3372           NaN         1700.0           20.0         0              NaN   

      trx_inkind_value_usd  trx_inkind_value_kes  trx_stdtime_mnths_hh_nr  \
3372                   NaN                   NaN                 11.55172   

      acc_unexplained_trx  
3372                  0.0

# Let's now only use the loans that have no unexplained balance adjustments.

In [104]:
dfb = dfb[dfb["acc_unexplained_trx"]==0]

# 2. Truncation into chunks with no new borrowing

We first insert a column for the new account numbers. These will be like the old account numbers but with a .1 or .2 for each chunk after a new borrowing.

In [105]:
# this is sometimes necessary if we mess smt up
# del dfb["new_account_ids"]

In [106]:
dfb.insert(df.columns.get_loc("account_ids")+1,"new_account_ids", ["-"]*(dfb.shape[0]))

### Now let's adjust this such that new borrowings are always the first thing on one day, and then run the same program as before.

In [110]:
accs = []
for acc in dfb["account_ids"].unique():
    accs.append(acc)
    i=0
    # we will use "indices" to save the reordered indices of the dataset for each unique loan
    indices = []
    # we will use "index_bor" to save the index of the new borrowing that happens on a given day,
    #whenever there were also other transactions (since we need that index to be first)
    index_bor = 0
    #only keep the rows of that account
    lna = dfb[dfb["account_ids"]==acc]
    # sort the dataset according to time
    lna = lna.sort_values("trx_stdtime_days_acc")
    # for each day in the dataset of the account, we check whether there was 
    # a new borrowing or starting balance and if yes reorder
    for d in lna["trx_stdtime_days_acc"].unique():
        # if only one entry that day we just add the index to indices
        if lna[lna["trx_stdtime_days_acc"]==d].shape[0]==1:
            indices.append(lna[lna["trx_stdtime_days_acc"]==d].index[0])
        # if there are more than 1 entry that day
        elif lna[lna["trx_stdtime_days_acc"]==d].shape[0]>1:
            # if there are more than one new borrowings or starting balances, we need to check manually
            if lna[(lna["trx_stdtime_days_acc"]==d)&((lna["trx_prx_purpose"]=="2. New borrowing")|(lna["trx_prx_purpose"]=="2. New borrowinng (individual portion)")|(lna["trx_prx_purpose"]=="1. Starting balance (today)"))].shape[0]>1:
                print("Several new borrowings on day " + str(d) +", check manually account " + acc + " in category "+ str(lna["trx_type_code"].unique()))
                for i in lna[lna["trx_stdtime_days_acc"]==d].index:
                            indices.append(i)
            #if there is exactly one new borrowing or starting balance that day (here we will need to check for each loan type what they exactly call new borrowings)
            elif lna[(lna["trx_stdtime_days_acc"]==d)&((lna["trx_prx_purpose"]=="2. New borrowing")|(lna["trx_prx_purpose"]=="2. New borrowinng (individual portion)")|(lna["trx_prx_purpose"]=="1. Starting balance (today)"))].shape[0]==1:
                #for each type of new borrowing if there is one that day, put that index first into indices
                for kind in ["2. New borrowing","2. New borrowinng (individual portion)","1. Starting balance (today)"]:
                    if kind in lna[lna["trx_stdtime_days_acc"]==d]["trx_prx_purpose"].unique():
                        index_bor = lna[(lna["trx_stdtime_days_acc"]==d) & (lna["trx_prx_purpose"]==kind)].index[0]
                        indices.append(index_bor)
                        #then add all other indices of that day
                        for i in lna[lna["trx_stdtime_days_acc"]==d].index.drop(index_bor):
                            indices.append(i)
            # if there are several entries for that day but none of them new borrowings just add all indices of that day            
            else:
                for i in lna[lna["trx_stdtime_days_acc"]==d].index:
                    indices.append(i)
    # here we reindex the lna dataset
    lna = lna.loc[indices,]
    #v will be the index of the chunk we are currently at
    v=0
    # saving the old account number
    acc = lna["account_ids"].unique()[0]
    # a is the new account number for each chunk
    a = lna["account_ids"].unique()[0]+".0"
    # we go through all rows of lna and change the entries of the column "new_account_ids" in the original dataset.
    for index, row in lna.iterrows():
        # if there is a new borrowing in the row, increase v by 1 and set a to acc.v
        if row["trx_prx_purpose"] == "1. Starting balance (today)" or row["trx_prx_purpose"] == "2. New borrowinng (individual portion)" or row["trx_prx_purpose"] == "2. New borrowing":
            v +=1
            a = acc+"."+str(v)
            dfb.at[index,"new_account_ids"]=a
        # otherwise set the new account number in that row to a
        else:
             dfb.at[index,"new_account_ids"]=a

Several new borrowings on day 266, check manually account 57137225366300000 in category [2752]
Several new borrowings on day 244, check manually account 61135946960100000 in category [3395]
Several new borrowings on day 0, check manually account 61134787763500000 in category [2752]


Several borrowings on the same day.
57137225366300000
61135946960100000
61134787763500000

In [118]:
df[df["account_ids"]=="57137225366300000"].sort_values("trx_stdtime_days_acc")[["trx_type_desc","trx_stdtime_days_acc","trx_prx_purpose","trx_value_kes","trx_bsheet_direction"]]

trx_type_desc  trx_stdtime_days_acc  \
476372  Friends and family: Borrowing                     0   
475250  Friends and family: Borrowing                     4   
473295  Friends and family: Borrowing                     6   
476508  Friends and family: Borrowing                    14   
475933  Friends and family: Borrowing                    38   
475125  Friends and family: Borrowing                   175   
475997  Friends and family: Borrowing                   182   
474926  Friends and family: Borrowing                   188   
475005  Friends and family: Borrowing                   203   
474162  Friends and family: Borrowing                   220   
473806  Friends and family: Borrowing                   234   
474308  Friends and family: Borrowing                   238   
474317  Friends and family: Borrowing                   239   
474516  Friends and family: Borrowing                   244   
474763  Friends and family: Borrowing                   245   
474302  Friends and family: Borrowing                   249   
476113  Friends and family: Borrowing                   258   
473329  Friends and family: Borrowing                   262   
474018  Friends and family: Borrowing                   266   
475197  Friends and family: Borrowing                   266   
474307  Friends and family: Borrowing                   279   
476358  Friends and family: Borrowing                   285   
475638  Friends and family: Borrowing                   297   
474059  Friends and family: Borrowing                   311   
233179  Friends and family: Borrowing                   316   
212904  Friends and family: Borrowing                   323   
473109  Friends and family: Borrowing                   339   
2382    Friends and family: Borrowing                   341   

                           trx_prx_purpose  trx_value_kes trx_bsheet_direction  
476372                    2. New borrowing         1000.0             Increase  
475250                    2. New borrowing         1000.0             Increase  
473295                         3. Payments         2000.0             Decrease  
476508                    2. New borrowing         1300.0             Increase  
475933                         3. Payments         1300.0             Decrease  
475125                    2. New borrowing          200.0             Increase  
475997                    2. New borrowing          500.0             Increase  
474926                         3. Payments          700.0             Decrease  
475005                    2. New borrowing          600.0             Increase  
474162                         3. Payments          600.0             Decrease  
473806                    2. New borrowing          200.0             Increase  
474308                    2. New borrowing         1000.0             Increase  
474317                    2. New borrowing         1000.0             Increase  
474516                         3. Payments         1000.0             Decrease  
474763                    2. New borrowing         1000.0             Increase  
474302                         3. Payments         2000.0             Decrease  
476113                    2. New borrowing          200.0             Increase  
473329                    2. New borrowing          200.0             Increase  
474018                    2. New borrowing          200.0             Increase  
475197                    2. New borrowing          500.0             Increase  
474307                         3. Payments         1100.0             Decrease  
476358                    2. New borrowing          450.0             Increase  
475638                    2. New borrowing          600.0             Increase  
474059                         3. Payments         1000.0             Decrease  
233179                    2. New borrowing          350.0             Increase  
212904                    2. New borrowing          350.0             Increase  
473109   

In [114]:
df[df["account_ids"]=="61134787763500000"][["trx_stdtime_days_acc","trx_prx_purpose","trx_value_kes","trx_bsheet_direction"]]

trx_stdtime_days_acc              trx_prx_purpose  trx_value_kes  \
17354                     0  1. Starting balance (today)            0.0   
22833                     0  1. Starting balance (today)            0.0   

      trx_bsheet_direction  
17354             Increase  
22833             Increase

The account 61135946960100000 seems to have several new borrowings on day 244.

In [115]:
df[df["account_ids"]=="61135946960100000"][["trx_stdtime_days_acc","trx_prx_purpose","trx_value_kes","trx_bsheet_direction"]]

trx_stdtime_days_acc                     trx_prx_purpose  \
5052                     285  7. Closing Balance--End of last DQ   
43429                    141                    2. New borrowing   
118905                     0                    2. New borrowing   
195059                     5                         3. Payments   
212855                   244                    2. New borrowing   
295268                   244                    2. New borrowing   

        trx_value_kes trx_bsheet_direction  
5052           3800.0                  NaN  
43429          2000.0             Increase  
118905          500.0             Increase  
195059          500.0             Decrease  
212855         1300.0             Increase  
295268          500.0             Increase

### Should we just get rid of these accounts?

In [111]:
dfb = dfb[(dfb["account_ids"]!="57137225366300000")&(dfb["account_ids"]!="61135946960100000")&(dfb["account_ids"]!="61134787763500000")]

## Let's check if it worked.

Let's see if all accounts got a new name.

In [119]:
dfb[dfb["new_account_ids"]=="-"]

Empty DataFrame
Columns: [hh_ids, unique_hhs, first_trx_date_hh, last_trx_date_hh, tot_hh_daysofobs, tot_hh_monthsofobs, interview_designation, int_date, int_month, int_year, int_yr_mo, first_int_date, account_ids, new_account_ids, unique_accnts, m_ids_owner, unique_hm_owner, account_bsheet_desig, account_startclose_balance, account_formal, account_liquid, first_trx_date_acc, last_trx_date_acc, tot_acc_daysofobs, tot_acc_monthsofobs, trx_id, m_ids_trx, trx_date, trx_month, trx_year, trx_yr_mo, trx_dq_round, trx_stdtime_days_hh, trx_stdtime_mnths_hh, trx_stdtime_days_acc, trx_stdtime_mnths_acc, trx_class_code, trx_class_desc, trx_family_code, trx_family_desc, trx_type_code, trx_type_desc, trx_prx_purpose, trx_prx_purpose_fd, trx_fee, trx_bsheet_direction, trx_mode_code, trx_mode_desc, trx_place_incommunity, trx_distance_km, trx_outlet, trx_direction, trx_value_kes, trx_value_usd, ddd_gift, trx_inkind_units, trx_inkind_value_usd, trx_inkind_value_kes, trx_stdtime_mnths_hh_nr, acc_unexplained_trx]
Index: []

### Let's look at the truncation of a few family and friends borrowings.

In [123]:
for i in range(0,20):
    acc = random.choice(dfb[(dfb["trx_type_code"]==2752)]["account_ids"].unique())
    display(dfb[dfb["account_ids"]==acc].sort_values("trx_stdtime_days_acc")[["trx_type_desc","trx_stdtime_days_acc","trx_value_kes","trx_prx_purpose","account_ids","new_account_ids"]])

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
397876  Friends and family: Borrowing                     0         3000.0   
158943  Friends and family: Borrowing                    16         3000.0   
3626    Friends and family: Borrowing                   176            0.0   

                           trx_prx_purpose        account_ids  \
397876                    2. New borrowing  57136617744800000   
158943                         3. Payments  57136617744800000   
3626    7. Closing Balance--End of last DQ  57136617744800000   

            new_account_ids  
397876  57136617744800000.1  
158943  57136617744800000.1  
3626    57136617744800000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
422313  Friends and family: Borrowing                     0         3000.0   
130218  Friends and family: Borrowing                    47         3000.0   
16888   Friends and family: Borrowing                   332            0.0   

                           trx_prx_purpose        account_ids  \
422313                    2. New borrowing  63135523008000000   
130218                         3. Payments  63135523008000000   
16888   7. Closing Balance--End of last DQ  63135523008000000   

            new_account_ids  
422313  63135523008000000.1  
130218  63135523008000000.1  
16888   63135523008000000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
158562  Friends and family: Borrowing                     0          300.0   
8560    Friends and family: Borrowing                    17          300.0   

                           trx_prx_purpose        account_ids  \
158562                    2. New borrowing  65137930163300000   
8560    7. Closing Balance--End of last DQ  65137930163300000   

            new_account_ids  
158562  65137930163300000.1  
8560    65137930163300000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
428866  Friends and family: Borrowing                     0        20000.0   
10682   Friends and family: Borrowing                    76        20000.0   

                           trx_prx_purpose         account_ids  \
428866                    2. New borrowing  108137416086400000   
10682   7. Closing Balance--End of last DQ  108137416086400000   

             new_account_ids  
428866  108137416086400000.1  
10682   108137416086400000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
3746    Friends and family: Borrowing                     0         4000.0   
359116  Friends and family: Borrowing                    30         4000.0   
18303   Friends and family: Borrowing                   358            0.0   

                           trx_prx_purpose        account_ids  \
3746           1. Starting balance (today)  60134725801400000   
359116                         3. Payments  60134725801400000   
18303   7. Closing Balance--End of last DQ  60134725801400000   

            new_account_ids  
3746    60134725801400000.1  
359116  60134725801400000.1  
18303   60134725801400000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
395877  Friends and family: Borrowing                     0         1000.0   
331530  Friends and family: Borrowing                     1          500.0   
128846  Friends and family: Borrowing                    28          500.0   
23022   Friends and family: Borrowing                    99            0.0   

                           trx_prx_purpose        account_ids  \
395877                    2. New borrowing  89137591184100000   
331530                         3. Payments  89137591184100000   
128846                         3. Payments  89137591184100000   
23022   7. Closing Balance--End of last DQ  89137591184100000   

            new_account_ids  
395877  89137591184100000.1  
331530  89137591184100000.1  
128846  89137591184100000.1  
23022   89137591184100000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
82284   Friends and family: Borrowing                     0          500.0   
119617  Friends and family: Borrowing                    46          150.0   
89663   Friends and family: Borrowing                    52           50.0   
261171  Friends and family: Borrowing                    58           50.0   
123155  Friends and family: Borrowing                    70           50.0   
315820  Friends and family: Borrowing                    70          100.0   
278626  Friends and family: Borrowing                    74           50.0   
53362   Friends and family: Borrowing                    78           50.0   
20897   Friends and family: Borrowing                   263            0.0   

                           trx_prx_purpose        account_ids  \
82284                     2. New borrowing  62136429071700000   
119617                         3. Payments  62136429071700000   
89663                          3. Payments  62136429071700000   
261171                         3. Payments  62136429071700000   
123155                         3. Payments  62136429071700000   
315820                         3. Payments  62136429071700000   
278626                         3. Payments  62136429071700000   
53362                          3. Payments  62136429071700000   
20897   7. Closing Balance--End of last DQ  62136429071700000   

            new_account_ids  
82284   62136429071700000.1  
119617  62136429071700000.1  
89663   62136429071700000.1  
261171  62136429071700000.1  
123155  62136429071700000.1  
315820  62136429071700000.1  
278626  62136429071700000.1  
53362   62136429071700000.1  
20897   62136429071700000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
5047    Friends and family: Borrowing                     0            0.0   
53097   Friends and family: Borrowing                   230          100.0   
119380  Friends and family: Borrowing                   276          100.0   
22      Friends and family: Borrowing                   381            0.0   

                           trx_prx_purpose        account_ids  \
5047           1. Starting balance (today)  60136798779200000   
53097                     2. New borrowing  60136798779200000   
119380                         3. Payments  60136798779200000   
22      7. Closing Balance--End of last DQ  60136798779200000   

            new_account_ids  
5047    60136798779200000.1  
53097   60136798779200000.2  
119380  60136798779200000.2  
22      60136798779200000.2

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
189998  Friends and family: Borrowing                     0         1000.0   
90606   Friends and family: Borrowing                    19          500.0   
396682  Friends and family: Borrowing                    24          500.0   
23660   Friends and family: Borrowing                   320            0.0   

                           trx_prx_purpose        account_ids  \
189998                    2. New borrowing  60135456285000000   
90606                          3. Payments  60135456285000000   
396682                         3. Payments  60135456285000000   
23660   7. Closing Balance--End of last DQ  60135456285000000   

            new_account_ids  
189998  60135456285000000.1  
90606   60135456285000000.1  
396682  60135456285000000.1  
23660   60135456285000000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
418353  Friends and family: Borrowing                     0        10000.0   
46619   Friends and family: Borrowing                     4        10000.0   
93773   Friends and family: Borrowing                   134            0.0   

                           trx_prx_purpose        account_ids  \
418353                    2. New borrowing  89137162126300000   
46619                          3. Payments  89137162126300000   
93773   7. Closing Balance--End of last DQ  89137162126300000   

            new_account_ids  
418353  89137162126300000.1  
46619   89137162126300000.1  
93773   89137162126300000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
480742  Friends and family: Borrowing                     0          400.0   
480738  Friends and family: Borrowing                    26          400.0   
21163   Friends and family: Borrowing                   310            0.0   

                           trx_prx_purpose        account_ids  \
480742                    2. New borrowing  57136272901000000   
480738                         3. Payments  57136272901000000   
21163   7. Closing Balance--End of last DQ  57136272901000000   

            new_account_ids  
480742  57136272901000000.1  
480738  57136272901000000.1  
21163   57136272901000000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
206542  Friends and family: Borrowing                     0          300.0   
143330  Friends and family: Borrowing                     8          400.0   
391222  Friends and family: Borrowing                    14          700.0   
42365   Friends and family: Borrowing                    93          110.0   
2892    Friends and family: Borrowing                    99          110.0   

                           trx_prx_purpose        account_ids  \
206542                    2. New borrowing  65137153487600000   
143330                    2. New borrowing  65137153487600000   
391222                         3. Payments  65137153487600000   
42365                     2. New borrowing  65137153487600000   
2892    7. Closing Balance--End of last DQ  65137153487600000   

            new_account_ids  
206542  65137153487600000.1  
143330  65137153487600000.2  
391222  65137153487600000.2  
42365   65137153487600000.3  
2892    65137153487600000.3

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
11736  Friends and family: Borrowing                     0         8000.0   

                   trx_prx_purpose        account_ids      new_account_ids  
11736  1. Starting balance (today)  61134770551400000  61134770551400000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
23821  Friends and family: Borrowing                     0            0.0   

                          trx_prx_purpose        account_ids  \
23821  7. Closing Balance--End of last DQ  50135237176900000   

           new_account_ids  
23821  50135237176900000.0

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
482886  Friends and family: Borrowing                     0          400.0   
482842  Friends and family: Borrowing                     6          400.0   
482771  Friends and family: Borrowing                     9         1000.0   
1911    Friends and family: Borrowing                   147         1000.0   

                           trx_prx_purpose        account_ids  \
482886                    2. New borrowing  57136827549200000   
482842                         3. Payments  57136827549200000   
482771                    2. New borrowing  57136827549200000   
1911    7. Closing Balance--End of last DQ  57136827549200000   

            new_account_ids  
482886  57136827549200000.1  
482842  57136827549200000.1  
482771  57136827549200000.2  
1911    57136827549200000.2

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
3175    Friends and family: Borrowing                     0          700.0   
186377  Friends and family: Borrowing                   218          700.0   
2439    Friends and family: Borrowing                   399            0.0   

                           trx_prx_purpose        account_ids  \
3175           1. Starting balance (today)  60136715715500000   
186377                         3. Payments  60136715715500000   
2439    7. Closing Balance--End of last DQ  60136715715500000   

            new_account_ids  
3175    60136715715500000.1  
186377  60136715715500000.1  
2439    60136715715500000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
27593   Friends and family: Borrowing                     0         2000.0   
134185  Friends and family: Borrowing                    27         2000.0   
19001   Friends and family: Borrowing                    46            0.0   

                           trx_prx_purpose        account_ids  \
27593                     2. New borrowing  89137995323700000   
134185                         3. Payments  89137995323700000   
19001   7. Closing Balance--End of last DQ  89137995323700000   

            new_account_ids  
27593   89137995323700000.1  
134185  89137995323700000.1  
19001   89137995323700000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
77754   Friends and family: Borrowing                     0          500.0   
321017  Friends and family: Borrowing                     7          500.0   
6193    Friends and family: Borrowing                    48            0.0   

                           trx_prx_purpose         account_ids  \
77754                     2. New borrowing  105137637052200000   
321017                         3. Payments  105137637052200000   
6193    7. Closing Balance--End of last DQ  105137637052200000   

             new_account_ids  
77754   105137637052200000.1  
321017  105137637052200000.1  
6193    105137637052200000.1

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
89166   Friends and family: Borrowing                     0          800.0   
483114  Friends and family: Borrowing                    77         2030.0   
152262  Friends and family: Borrowing                   147         1200.0   
337685  Friends and family: Borrowing                   254          600.0   
307257  Friends and family: Borrowing                   301          600.0   
352156  Friends and family: Borrowing                   396           30.0   
2602    Friends and family: Borrowing                   406            0.0   

                           trx_prx_purpose        account_ids  \
89166                          3. Payments  62134702499000000   
483114                    2. New borrowing  62134702499000000   
152262                         3. Payments  62134702499000000   
337685                    2. New borrowing  62134702499000000   
307257                         3. Payments  62134702499000000   
352156                         3. Payments  62134702499000000   
2602    7. Closing Balance--End of last DQ  62134702499000000   

            new_account_ids  
89166   62134702499000000.0  
483114  62134702499000000.1  
152262  62134702499000000.1  
337685  62134702499000000.2  
307257  62134702499000000.2  
352156  62134702499000000.2  
2602    62134702499000000.2

trx_type_desc  trx_stdtime_days_acc  trx_value_kes  \
439986  Friends and family: Borrowing                     0         1027.0   
440025  Friends and family: Borrowing                    11         1027.0   
11353   Friends and family: Borrowing                    97            0.0   

                           trx_prx_purpose        account_ids  \
439986                    2. New borrowing  89137578416400000   
440025                         3. Payments  89137578416400000   
11353   7. Closing Balance--End of last DQ  89137578416400000   

            new_account_ids  
439986  89137578416400000.1  
440025  89137578416400000.1  
11353   89137578416400000.1

In [124]:
dfb[dfb["account_ids"]=="112137283351600000"]

hh_ids  unique_hhs first_trx_date_hh last_trx_date_hh  \
252     KMOMK03         NaN         07sep2012        25oct2013   
20371   KMOMK03         NaN         07sep2012        25oct2013   
236977  KMOMK03         NaN         07sep2012        25oct2013   
320008  KMOMK03         NaN         07sep2012        25oct2013   
443714  KMOMK03         NaN         07sep2012        25oct2013   

        tot_hh_daysofobs  tot_hh_monthsofobs interview_designation   int_date  \
252                  413                  14  04=Diaries Interview  10oct2013   
20371                413                  14  04=Diaries Interview  03jul2013   
236977               413                  14  04=Diaries Interview  19jul2013   
320008               413                  14  04=Diaries Interview  29aug2013   
443714               413                  14  04=Diaries Interview  03jul2013   

        int_month  int_year int_yr_mo first_int_date         account_ids  \
252            10      2013   2013_10      01sep2012  112137283351600000   
20371           7      2013   2013_07      01sep2012  112137283351600000   
236977          7      2013   2013_07      01sep2012  112137283351600000   
320008          8      2013   2013_08      01sep2012  112137283351600000   
443714          7      2013   2013_07      01sep2012  112137283351600000   

             new_account_ids  unique_accnts        m_ids_owner  \
252     112137283351600000.1            NaN  61134406652800000   
20371   112137283351600000.1            1.0  61134406652800000   
236977  112137283351600000.1            NaN  61134406652800000   
320008  112137283351600000.1            NaN  61134406652800000   
443714  112137283351600000.1            NaN  61134406652800000   

        unique_hm_owner account_bsheet_desig account_startclose_balance  \
252                 NaN            Liability                      Close   
20371               NaN            Liability                        NaN   
236977              NaN            Liability                        NaN   
320008              NaN            Liability                        NaN   
443714              NaN            Liability                        NaN   

       account_formal account_liquid first_trx_date_acc last_trx_date_acc  \
252            Formal            NaN          20jun2013         10oct2013   
20371          Formal            NaN          20jun2013         10oct2013   
236977         Formal            NaN          20jun2013         10oct2013   
320008         Formal            NaN          20jun2013         10oct2013   
443714         Formal            NaN          20jun2013         10oct2013   

        tot_acc_daysofobs  tot_acc_monthsofobs              trx_id  \
252                   112                    3  112138138546700000   
20371                 112                    3  112137283415200000   
236977                112                    3  112137431148400000   
320008                112                    3  112137831140900000   
443714                112                    3  112137283404000000   

                m_ids_trx   trx_date  trx_month  trx_year trx_yr_mo  \
252     61134406652800000  10oct2013         10      2013   2013_10   
20371   61134406652800000  20jun2013          6      2013   2013_06   
236977  61134406652800000  15jul2013          7      2013   2013_07   
320008  61134406652800000  26aug2013          8      2013   2013_08   
443714  61134406652800000  20jun2013          6      2013   2013_06   

        trx_dq_round  trx_stdtime_days_hh  trx_stdtime_mnths_hh  \
252             19.0                  398                    13   
20371           13.0                  286                     9   
236977          14.0                  311                    10   
320008          17.0                  353                    12   
443714          13.0                  286                     9   

        trx_stdtime_days_acc  trx_stdtime_mnths_acc trx_class_code  \
252                      112        

In [125]:
dfb.to_csv("diaries_trx_trunc_loans.csv")

### Careful: the column "unique_accnts" is not accurate anymore.